modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -146,6 +154,74 @@
+;; https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#floating_point_types
+(def ^:private dec-fingerprint-types
+  "Decimal and floating point types for fingerprinting purposes (to be parsed via BigDecimal)"
+  #{"NUMERIC" "DECIMAL" "BIGNUMERIC" "BIGDEIMAL" "FLOAT" "FLOAT64"})
+
+(defn- FieldValueList->fingerprint-structure [fields ^FieldValueList values]
+  (map-indexed (fn [idx ^FieldValue fv]
+                 (let [{database-type :database_type} (nth fields idx)
+                       v                              (.getValue fv)]
+                      (if (string? v)
+                        (condp contains? (some-> database-type str/upper-case)
+                          #{"TIMESTAMP"} (bq-ts-str->instant v)
+                          int-fingerprint-types (java.math.BigInteger. ^String v)
+                          dec-fingerprint-types (java.math.BigDecimal. ^String v)
+                          v)
+                        v)))
+               ;; the list operation will return *all* field values for each row, in the database field order
+               ;; so, map over the requested fields and fetch only those values at each respective :database_position
+               (map (fn [{^Integer database-pos :database_position}]
+                      (.get values database-pos))
+                    fields)))
+
+;; Override the table-rows-sample multimethod for :bigquery-cloud-sdk, to use a different mechanism for sampling
+;; table rows (the list operations, rather than a regular query), purely as a cost control measure.  For certain types
+;; of tables, BigQuery evidently does not honor the limit parameter of a query, and instead performs a full table scan
+;; and then truncates results at the end.  See #19470 for full details.
+(defmethod @#'sync.f/table-rows-sample :bigquery-cloud-sdk [driver
+                                                            {table-name :name, dataset-id :schema :as table}
+                                                            fields
+                                                            rff
+                                                            opts]
+  (let [database (table/database table)
+        bq-tbl   (get-table database dataset-id table-name)
+        tbl-type (-> (.getDefinition bq-tbl)
+                     (.getType))]
+    (if (contains? #{TableDefinition$Type/VIEW TableDefinition$Type/MATERIALIZED_VIEW} tbl-type)
+      (do (log/debugf "%s.%s is a view, so we cannot use the list API; falling back to regular query"
+                      dataset-id
+                      table-name)
+          ((get-method @#'sync.f/table-rows-sample :sql-jdbc) driver table fields rff opts))
+      (let [^TableResult rows (.list bq-tbl (u/varargs BigQuery$TableDataListOption))
+            num-rows          metadata-queries/max-sample-rows]
+        (transduce (comp (take num-rows)
+                         (map (partial FieldValueList->fingerprint-structure fields)))
```

Comment:
One thing i'm noticing, should we replicate how we do the `row-thunks` in the normal query processor? I'm looking at the `FieldValueList->fingerprint-structure` function. I see that a `TableResult` has a `getSchema` function. Perhaps we should grab the schema, toss that into a function that will return a row thunk function? With the `partial`ed version here we keep checking how to parse each row and looking up indices we care about, when we could do this once and then return that once, and reuse it on each row.

Reply:
I seem to recall that this API doesn't offer the `getSchema` from `TableResult`.  Like it throws an `UnsupportedOperationException` or some such.  But I'll double check.  In any case, there is probably a better way to build the parsers only once from `fields` as you suggested.